### ISSUE TITLE
Profiling Filter Tool Ideas

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos-tools#88]
This is a collection of ideas for setting the profiling filter via expressions

```c++
// start and stop based on entering region at a certain iteration
// no :ITERATION given start stop and any entry/exit
global_start [ region REGEX : ITERATION ]
global_stop [ region REGEX : ITERATION ]

// Specific Kernel Always
REGEX

// Specific Kernel Regex inside region
// ITERATION_START/STOP optional, only one number: only profile in that region instance
REGEX [ region REGION_REGEX : ITERATION_START : ITERATION_STOP ]

// Specific Kernel only if as parallel_reduce in two specific nested regions
REGEX [ type parallel_reduce ] [ region REGION_REGEX1 ] [ region REGION_REGEX2 ]

// Specific Kernel but only 1011 call
REGEX:1011

// Specific Kernel but only after sequence of other kernels
REGEX [ sequence REGEX1 REGEX2 REGEX3 ]
```

**Comment by DavidPoliakoff** [comment]
Additions via Slack

```
load_tool [name SOME_LIBRARY.so]
```

Says to load SOME_LIBRARY.so and associate it with name

Then any of the REGEX expressions above can have a set of tools added, for example

```
REGEX [ type parallel_reduce ] [ region REGION_REGEX1 ] [ region REGION_REGEX2 ] tool TOOL1 tool TOOL2
```

Which will forward things matching that to TOOL1 and TOOL2

**Comment by mhoemmen** [comment]
* I like "Specific Kernel Regex inside region"
* "Specific Kernel but only after sequence of other kernels" sounds like you'll need to be careful to bound the history window, else the profiling library will eat up memory

**Comment by crtrott** [comment]
Other thing fencing:
```
global_fencing off

REGEX [ fence off ]
```

**Comment by crtrott** [comment]
@mhoemmen yeah and the search time becomes too large. I would just say we limit to like a sequence of 5, and essentially keep a revolving list of the last 5 kernel calls or so.

**Comment by DavidPoliakoff** [comment]
>     * I like "Specific Kernel Regex inside region"
> 
>     * "Specific Kernel but only after sequence of other kernels" sounds like you'll need to be careful to bound the history window, else the profiling library will eat up memory

Memory exists for tools, codes should stop being so greedy... :p

**Comment by mhoemmen** [comment]
> I would just say we limit to like a sequence of 5, and essentially keep a revolving list of the last 5 kernel calls or so.

The list of filters themselves should tell you how much space you need, but if y'all start building crazy state machines that eat memory, this could be an issue.

**Comment by crtrott** [comment]
the good thing is that you only pay for it if you use it ...
